Added onedirectional 2D-LSTM

ChristopherBrix opened the pull request:
I've added support for a onedirectional 2D-LSTM in TensorFlow.

The code is based on the Theano implementation for 4 directions, and is not as general as I would like. But it might be helpful nevertheless.

The inputs to the 2D-LSTM layer have to be two sequences (eg. source and target). The output is a sequence, not a 2D grid, so it can easily be combined with other layers.

During inference, it is assumed that the shape of the second (target) dimension is always 1 (the previous target word).

This might be relevant for issue #8.

albertz commented:
Can you rebase and resolve the conflicts? (Also remove the merge commit. Just rebase it.)

albertz commented on TFNetwork.py:
```diff
@@ -1120,7 +1120,8 @@ def get_rec_step_info(self, must_exist=True):
     rec_layer = self.get_rec_parent_layer()
-    if not rec_layer:
+    # the second condition is true if all layers have been optimized out of the rec layer
+    if not rec_layer or len(rec_layer.cell.layers_in_loop) == 0:
```
I don't understand.
If `rec_layer` is None, the code below will lead to an exception (any code below which access `rec_layer`).
Also, `must_exist` specifically says that it must exists. So if it does not exist, it should throw an exception (that is what `must_exists` is for).

ChristopherBrix replied:
If `rec_layer` is None, the first condition is true. Then, it is ensured that `must_exist` is false, otherwise an exception is thrown as expected. If `must_exist` is false, None is returned, indicating that there is no recurrent step information. This behavior is used by `have_rec_step_info()`.

If `rec_layer` is not None, but all contained layers have been optimized out of the loop, my modification ensures that no recurrent step information is returned. This is necessary, because it was never created (see https://github.com/rwth-i6/returnn/blob/master/TFNetworkRecLayer.py#L1580).

If it is not None and contains internal layers, then the step information is returned as expected.

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -4383,3 +4383,273 @@ def __call__(self, inputs, state, scope=None):
 
     return new_c, new_h
+
+class TwoDLSTMLayer(LayerBase):
```
Two empty lines. See my earlier comment.

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -4383,3 +4383,273 @@ def __call__(self, inputs, state, scope=None):
+      self._add_params(params=params, scope_name_prefix=scope_name_prefix)
+
+  @classmethod
+  def _post_init_output(cls, output, network, target=None, size_target=None, sources=(), **kwargs):
```
Why do you need this function? Everything what you do here should be done in `__init__`, if possible.

ChristopherBrix replied:
Don't I have to overwrite this method, as explained in the [`LayerBase` class](https://github.com/rwth-i6/returnn/blob/master/TFNetworkLayer.py#L256)?

However, moving it to  `__init__` would have the benefit that I can directly check `have_rec_step_info` and don't have to rely on `train_flag`. For my use case, both are equivalent, but they don't necessary have to be.

albertz replied:
No, it should only be used if not possible via `__init__` (which is very rarely the case).

albertz replied:
This still need to be resolved.

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -4383,3 +4383,273 @@ def __call__(self, inputs, state, scope=None):
+      n_hidden=n_hidden, n_input_dim=sum(n_input_dim_parts), n_input_dim_parts=n_input_dim_parts,
+      input_is_sparse=self.sources[0].output.sparse,
+      **unit_opts,
+      pooling=self.pooling)
```
I think this is Python 3 syntax, right? The code should also work (or at least load without syntax error) in Python 2 (at least for a while).

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -4383,3 +4383,273 @@ def __call__(self, inputs, state, scope=None):
+        mark_data_key_as_used=network.train_flag is not False).size_placeholder.copy()
+    else:
+      with tf.control_dependencies(None):
+        output.size_placeholder = {} # Not sure why this has to be defined. Is it ever used?
```
Follow Python code style conventions, ie. two spaces before `#`.
Also, remove the comment anyway here. The `output.size_placeholder` is used in many places. (Let's discuss offline if you wonder where...)

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -4383,3 +4383,273 @@ def __call__(self, inputs, state, scope=None):
+        self.output.batch_dim_axis = 0
+      else:
+        self.output.time_dim_axis = 0
+        self.output.batch_dim_axis = 1
```
The layer output format should not be defined in `__init__`. This should be defined in `get_out_data_from_opts`, which you need to overload. See other layers for example.

ChristopherBrix replied:
The output format depends on whether or not the 2D layer was optimized out of the recurrent layer. Currently, I determine this by checking `self.network.have_rec_step_info()`. 

In the class method `get_out_data_from_opts`, the `network` doesn't appear to be available - how can I perform the check in there?

albertz replied:
This still needs to be resolved.

ChristopherBrix commented on NativeOp.cpp:
```diff
@@ -190,6 +190,78 @@ static void tf_cuda_sgemm(
+#else  // GOOGLE_CUDA
+    context->SetStatus(errors::InvalidArgument("CuBlasGemm needs CUDA."));
+#endif  // GOOGLE_CUDA
+}
```
`sgemm_batched` is needed for performance reasons. Usually, only one matrix matrix multiplication is needed at any time. But for the 2D-LSTM, we can compute a complete diagonal at once, so multiple operations can be performed in parallel. 

For CPU support, this could be changed to a sequence of multiple `sgemm` calls.

albertz replied:
I don't exactly understand. What is the matrix in `a_ptrs`, `b_ptr`, `c_ptr`, or better, what are their shapes?

ChristopherBrix replied:
`a_ptrs` (as well as `b_ptrs` and ` c_ptrs`) wraps the input `in_x` (`in_y`, `out`).
This input is a sequence of matrices (for 2D-LSTM, one matrix for each element of the current diagonal). The number of these matrices is given by `batch_size` (which is equal to `batchCount`, I'll make it consistent).
In the `for` loop, they are wrapped using `AsDeviceMemory()` and added to `a_ptrs`.
The shape of the matrices is given by `m`, `n` and `k`: `in_x` has shape `(m, k)`, `in_y` has shape `(k, n)` and `out` has shape `(m, n)`.

albertz replied:
So, (m,k), (k,n) are the input shapes, (m,n) is the output shape. This is a pretty standard matrix multiplication, which should be possible via normal `sgemm`. I don't see what `sgemm_batched` can do better/faster than `sgemm`.

ChristopherBrix replied:
But `sgemm` only performs one such multiplication at a time, right? We have to perform *multiple, independent* multiplications in parallel.

ChristopherBrix commented:
I don't know how much work will be required to add CPU support for 2D-LSTMs and when I'll have the time to implement it.

Maybe we can postpone this?

albertz commented:
Sure, as I said in my comment, let's skip the CPU for now.

albertz commented:
Btw, also see that the tests all run through (here on GitHub, there is the Travis link to your pull request).

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -4460,3 +4460,283 @@ def __call__(self, inputs, state, scope=None):
+      self._add_params(params=params, scope_name_prefix=scope_name_prefix)
+
+  @classmethod
+  def get_out_data_from_opts(cls, sources, **kwargs):
```
Why do you have only `sources` here, but not other kwargs which you use below, like `network`, etc?

albertz replied:
It also seems as if you assume that `len(sources) == 2`, right? If so, you should add such an `assert`.

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -4460,3 +4460,283 @@ def __call__(self, inputs, state, scope=None):
+      self._rec_scope = scope
+      scope_name_prefix = scope.name + "/"  # e.g. "layer1/rec/"
+      with self.var_creation_scope():
+        print("============!!!!!!!!!!!!========= Cell created")
```
Better remove those debug msgs.
If you really want them, forward them to the log stream, e.g. `file=log.v4`.
But please remove the `============!!!!!!!!!!!!========= `, and make the message more meaningful.

ChristopherBrix replied:
Sorry, I added a lot of debug messages while looking for the `size_placeholder` bug. Apparently, I forgot to delete this one afterwards.

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -4460,3 +4460,283 @@ def __call__(self, inputs, state, scope=None):
+      shape = (None, kwargs["n_out"],)
+      time_dim_axis = 0
+      batch_dim_axis = 1
+      size_placeholder = sources[1].output.size_placeholder.copy()
```
Do not add workarounds. Esp not such fragile workarounds which rely on some specific random behavior which could maybe change at some future point.
If there is a problem, the problem itself should be fixed, or if there is no good way to get some information, just introduce that. If needed. In this cases, it should not really be needed.
As I said, look just at how other layers do that, which is much simpler.
E.g. `SelfAttentionLayer`, or `RnnCellLayer`. Or `WindowLayer`, or `CumsumLayer`.
I guess the output shape is anyway very simple, just in all cases like `sources[1].output.shape`, or not?
Or you could check for `sources[1].output.time_dim_axis is None` or so.

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -4460,3 +4460,283 @@ def __call__(self, inputs, state, scope=None):
+    if is_gpu_available():
+      from tensorflow.contrib import cudnn_rnn
+    else:
+      assert(False) # currently, there's no CPU support
```
The proper syntax is `assert False`.
And the comment you added could just go into the assert message, which is more helpful for the user then. I.e. 
```suggestion
      assert False, "currently, there's no CPU support"
```

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -4460,3 +4460,283 @@ def __call__(self, inputs, state, scope=None):
+    # we have to specifiy them (and they must have 4 dims), because during inference (when multiple iterations
+    # for one 2DLSTM happen, as opposed to the training where everything is computed at once) we pass the intermediate
+    # values from one iteration to the other. 42 is added to show from the value of the variable that it's not
+    # supposed to be used.
```
The values are used. They are the initial values.
In case they are not needed (training), do not set them.
Again, see other layers as example, e.g. `SelfAttentionLayer`.

albertz commented on TFNetworkRecLayer.py:
```diff
@@ -4460,3 +4460,283 @@ def __call__(self, inputs, state, scope=None):
+    """
+    return {"state": tf.TensorShape((None, None, None, None)),
+            "output": tf.TensorShape((None, None, None, None)),
+            "iteration": tf.TensorShape((None,))}
```
See above.

ChristopherBrix commented:
I've fixed the issues you pointed out - is there anything else?
Also, in the meantime I've added two small improvements in my codebase: the ability to pick the maximum instead of the last state on the horizontal axis and to use dropout. Would you like me to add it to this PR, or to create a new one after this one is merged?

albertz commented:
Remove merge commits. Rebase on the latest master commit.
Also `get_rec_initial_extra_outputs` still looks wrong. If it is outside the rec layer, there should not be a recurrent state.

albertz commented:
Any new things, once this is merged in, I don't think it needs another PR, unless you are unsure maybe about the change.

albertz commented:
Hey, what's the state here?

albertz commented:
Hey, what's the state here? Also, can you rebase? (It conflicts currently to the master.)

ChristopherBrix commented:
Sorry, I didn't work on this for some time.
I've updated the `get_rec_initial_extra_outputs` method, now it returns an empty dict if there's no time dimension to loop over.

There's just one test failing, and the same tests fails on the master branch as well, so all should be fine.
